src/benchmark.ts

```diff
@@ -1,19 +1,139 @@
-export abstract class Benchmark {
-  startTime: number;
+import { errorToObject } from "./util";
 
-  constructor(private name: string) {}
+const DEFAULT_CONFIG: any = {
+  timeout: 10000
+};
 
-  async start() {
-    console.log(`Starting benchmark: ${this.name}`);
-    this.startTime = Date.now();
-  }
-
-  abstract async run(): Promise<void>;
+export abstract class Benchmark<
+  C extends BenchmarkConfig,
+  R extends BenchmarkReport
+> {
+  protected aborted = false;
+  protected config: C;
+  protected report: R = { execution: {}, result: {} } as R;
 
-  async end() {
-    console.log(
-      `End benchmark: ${this.name}. Total time: ${Date.now() -
-        this.startTime} ms`
+  constructor(config: C, defaultConfig: any) {
+    this.config = Object.assign(
+      {},
+      Object.assign(DEFAULT_CONFIG, defaultConfig),
+      config
     );
+    if (!this.config.name) {
```

Comment:
I don't think it's possible here because ellipsis "..." operator is used to combine multiple objects into this.config and then checks are applied to result... and I am not sure it would protect from passing empty strings